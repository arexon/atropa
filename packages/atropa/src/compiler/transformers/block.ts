import { deepMerge, objectMap } from '@antfu/utils'
import { ensureNamespaces } from '../utils'

interface UserTemplate {
	namespace?: string
	description?: (template: Record<string, any>) => void
	permutations?: (template: Record<string, any>[]) => void
	components?: (template: Record<string, any>) => void
	events?: (template: Record<string, any>) => void
	use?: (...components: Record<string, any>[]) => void
}
interface VanillaTemplate {
	description?: Record<string, any>
	permutations?: Record<string, any>[]
	components?: Record<string, any>
	events?: Record<string, any>
}
interface BlockObject {
	format_version: string
	'minecraft:block': VanillaTemplate
}

export function processTemplate(template: VanillaTemplate): UserTemplate {
	return {
		namespace: process._namespace,
		description: (_template) => {
			template.description = { ...template.description, ..._template }
		},
		components: (_template) => {
			template.components = { ...template.components, ..._template }
		},
		permutations: (_template) => {
			template.permutations = [
				...(template.permutations || []),
				..._template
			]
		},
		events: (_template) => {
			template.events = { ...template.events, ..._template }
		},
		use: (...components) => {
			deepMerge(template, ...components)
		}
	}
}

export function transformTemplate(
	template: VanillaTemplate,
	version: string
): BlockObject {
	const transformedTemplate = objectMap(
		template as Required<VanillaTemplate>,
		(key, value) => {
			if (key === 'components') {
				return [key, ensureNamespaces(value, 'minecraft')]
			}
			if (key === 'permutations' && Array.isArray(value)) {
				for (const permutation of value) {
					permutation.components = ensureNamespaces(
						permutation.components,
						'minecraft'
					)
				}
			}
			return [key, value]
		}
	) as VanillaTemplate

	return {
		format_version: version,
		'minecraft:block': transformedTemplate
	}
}
