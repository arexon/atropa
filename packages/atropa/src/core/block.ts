import {
	processTemplate,
	transformTemplate
} from '../compiler/transformers/block'
import type {
	BlockTemplate,
	BlockFormatVersion
} from '../schema/atropa/server/block'

/**
 * # Define Block
 *
 * Generates a new block based on the given templates.
 * @param version The format version of the block.
 * @param fn A callback function with function parameters used to define the block.
 * @returns A block.
 */
export function defineBlock<Version extends BlockFormatVersion>(
	version: Version,
	fn: (template: BlockTemplate<Version>) => void
): Record<string, any> {
	try {
		const template = {}

		fn(processTemplate(template) as BlockTemplate<Version>)
		return transformTemplate(template, version)
	} catch (error) {
		throw new Error(`Failed to transform block template`, error as Error)
	}
}
