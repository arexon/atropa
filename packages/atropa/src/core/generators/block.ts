import { deepMerge, objectMap } from '@antfu/utils'
import { ensureNamespaces } from '../utils'
import type {
	BlockTemplate,
	BlockFormatVersion
} from '../../schema/atropa/server/block'

interface UserTemplate {
	namespace?: string
	description?: (template: Record<string, any>) => void
	permutations?: (template: Record<string, any>[]) => void
	components?: (template: Record<string, any>) => void
	events?: (template: Record<string, any>) => void
	use?: (...components: Record<string, any>[]) => void
}
interface VanillaTemplate {
	description?: Record<string, any>
	permutations?: Record<string, any>[]
	components?: Record<string, any>
	events?: Record<string, any>
}
interface Block {
	format_version: string
	'minecraft:block': VanillaTemplate
}

/**
 * # Define Block
 *
 * Generates a new block based on the given templates.
 * @param version The format version of the block.
 * @param fn A callback function with function parameters used to define the block.
 * @returns A block.
 */
export function defineBlock<Version extends BlockFormatVersion>(
	version: Version,
	fn: (template: BlockTemplate<Version>) => void
): Record<string, any> {
	try {
		const template = {}

		fn(processTemplate(template) as BlockTemplate<Version>)
		return transformTemplate(template, version)
	} catch (error) {
		throw new Error(`Failed to transform block template`, error as Error)
	}
}

export function processTemplate(template: VanillaTemplate): UserTemplate {
	return {
		namespace: process._namespace,
		description: (_template) => {
			template.description = { ...template.description, ..._template }
		},
		components: (_template) => {
			template.components = { ...template.components, ..._template }
		},
		permutations: (_template) => {
			template.permutations = [
				...(template.permutations || []),
				..._template
			]
		},
		events: (_template) => {
			template.events = { ...template.events, ..._template }
		},
		use: (...components) => {
			deepMerge(template, ...components)
		}
	}
}

function transformTemplate(template: VanillaTemplate, version: string): Block {
	const transformedTemplate = objectMap(
		template as Required<VanillaTemplate>,
		(key, value) => {
			if (key === 'components') {
				return [key, ensureNamespaces(value, 'minecraft')]
			}
			if (key === 'permutations' && Array.isArray(value)) {
				for (const permutation of value) {
					permutation.components = ensureNamespaces(
						permutation.components,
						'minecraft'
					)
				}
			}
			return [key, value]
		}
	) as VanillaTemplate

	return {
		format_version: version,
		'minecraft:block': transformedTemplate
	}
}
