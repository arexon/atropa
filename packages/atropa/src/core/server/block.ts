import { deepMerge, objectMap } from '@antfu/utils'
import { ensureNamespaces, removeEmptyProperties } from '../utils'
import type {
	BlockTemplate,
	BlockFormatVersion
} from '../../schema/atropa/server/block'

type Object = Record<string, any>
interface InputTemplate {
	namespace?: string
	description?: (template: Object) => void
	permutations?: (template: Object[]) => void
	components?: (template: Object) => void
	events?: (template: Object) => void
	use?: (...components: Object[]) => void
}
interface OutputTemplate {
	description?: Object
	permutations?: Object[]
	components?: Object
	events?: Object
	use?: Object[]
}

/**
 * # Define Block
 *
 * Generates a new block based on the given templates.
 * @param version The format version of the block.
 * @param fn A callback function with function parameters used to define the block.
 * @returns A block.
 */
export function defineBlock<Version extends BlockFormatVersion>(
	version: Version,
	fn: (template: BlockTemplate<Version>) => void
): Record<string, any> {
	try {
		const isLegacy = version === '1.16.0'
		const template: OutputTemplate = {}

		fn(processTemplate(template, isLegacy) as BlockTemplate<Version>)

		return {
			format_version: version,
			'minecraft:block': transformTemplate(template, isLegacy)
		}
	} catch (error) {
		throw new Error(`Failed to parse block template`, error as Error)
	}
}

export function processTemplate(
	fields: OutputTemplate,
	isLegacy: boolean
): Object {
	const template: InputTemplate = {
		namespace: global.config.namespace,
		description: (template) => {
			fields.description = { ...fields.description, ...template }
		},
		components: (template) => {
			fields.components = { ...fields.components, ...template }
		},
		use: (...components) => {
			fields.use = [...(fields.use || []), ...components]
		}
	}

	if (!isLegacy) {
		deepMerge(template, {
			permutations: (template) => {
				fields.permutations = fields.permutations
					? [...fields.permutations, ...template]
					: template
			},
			events: (template) => {
				fields.events = { ...fields.events, ...template }
			}
		})
	}

	return template
}

function transformTemplate(fields: OutputTemplate, isLegacy: boolean): Object {
	const template: Object = {
		description: fields.description,
		components: fields.components
	}

	if (!isLegacy) {
		deepMerge(template, {
			permutations: fields.permutations,
			events: fields.events
		})
	}
	if (fields.use) deepMerge(template, ...fields.use)

	return objectMap(removeEmptyProperties(template), (key, value) => {
		if (key === 'components') {
			return [key, ensureNamespaces(value, 'minecraft')]
		}
		if (key === 'permutations') {
			for (const permutation of value) {
				if (permutation.components !== undefined) {
					permutation.components = ensureNamespaces(
						permutation.components,
						'minecraft'
					)
				}
			}
		}
		return [key, value]
	})
}
