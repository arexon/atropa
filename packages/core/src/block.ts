import { loadConfig } from 'volars-config'
import { BlockTemplate, FormatVersion } from 'volars-schema'
import { prependNamespacesInArray, prependNamespacesInObject } from './utils'
import { logger } from './logger'

/**
 * # Define Block
 *
 * Generates a new block based on the given templates.
 */
export async function defineBlock<Version extends FormatVersion>(
	formatVersion: Version,
	block: (template: BlockTemplate<Version>) => void
): Promise<Object | unknown> {
	type BlockObject = Record<string, any>
	const config = await loadConfig()

	let description: BlockObject = {}
	let components: BlockObject = {}
	let permutations: BlockObject[] = []

	try {
		block({
			namespace: config.namespace,
			description: (template: BlockObject) => (description = template),
			...(formatVersion !== '1.16.0' && {
				permutations: (template: BlockObject[]) => {
					permutations = template
				}
			}),
			components: (template: BlockObject) => (components = template)
		} as any)

		return {
			format_version: formatVersion,
			'minecraft:block': {
				description,
				...(formatVersion !== '1.16.0' && {
					permutations: prependNamespacesInArray(
						permutations,
						'components',
						'minecraft'
					)
				}),
				components: prependNamespacesInObject(components, 'minecraft')
			}
		}
	} catch (error) {
		logger.error(`Failed to parse block:`, error)
		process.exit(0)
	}
}
