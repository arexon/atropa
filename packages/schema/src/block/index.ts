import { FormatVersion } from '../FormatVersion'
import { Template_1_16_0 } from './v1.16.0'
import { Template_1_16_100 } from './v1.16.100'
import { Template_1_18_10 } from './v1.18.10'
import { Template_1_18_30 } from './v1.18.30'
import { Template_1_19_10 } from './v1.19.10'

export type BlockTemplate<Version extends FormatVersion> =
	Version extends '1.16.0'
		? Template_1_16_0
		: Version extends '1.16.100'
		? Template_1_16_100
		: Version extends '1.18.10'
		? Template_1_18_10
		: Version extends '1.18.30'
		? Template_1_18_30
		: Version extends '1.19.10'
		? Template_1_19_10
		: never
