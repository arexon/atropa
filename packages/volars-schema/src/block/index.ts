import { FormatVersion } from '../FormatVersion'
import { Description, Properties } from './Description'
import { Components_1_16_0 } from './v1.16.0'
import { Components_1_16_100 } from './v1.16.100'
import { Components_1_18_10 } from './v1.18.10'
import { Components_1_18_30 } from './v1.18.30'
import { Components_1_19_10 } from './v1.19.10'

export interface BlockTemplate<Version extends FormatVersion> {
	/**
	 * # Namespace
	 *
	 * The project namespace defined in `config.json`.
	 */
	namespace: string

	/**
	 * # Description
	 *
	 * The description sets required block information.
	 */
	description: (
		template: Version extends '1.16.0'
			? Description
			: Description & Properties
	) => void

	/**
	 * # Compnoents
	 *
	 * Components are used to describe the block's attributes and behavior.
	 */
	components: (
		template: Version extends '1.16.0'
			? Components_1_16_0
			: Version extends '1.16.100'
			? Components_1_16_100
			: Version extends '1.18.10'
			? Components_1_18_10
			: Version extends '1.18.30'
			? Components_1_18_30
			: Version extends '1.19.10'
			? Components_1_19_10
			: never
	) => void
}
