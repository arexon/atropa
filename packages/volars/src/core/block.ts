import { loadConfig } from '../config'

export async function defineBlock(
	fn: (template: BlockTemplate) => void
): Promise<Object> {
	const config = await loadConfig()

	let formatVersion = ''
	let description = {}
	let components = {}

	fn({
		namespace: config.namespace,
		formatVersion: (template) => (formatVersion = template),
		description: (template) => (description = template),
		components: (template) => (components = template)
	})

	return {
		format_version: formatVersion,
		'minecraft:block': {
			description,
			components
		}
	}
}

interface BlockTemplate {
	namespace?: string
	formatVersion: (template: string) => void
	description: (template: Object) => void
	components: (template: Object) => void
}
