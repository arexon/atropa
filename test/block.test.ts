import { describe, expect, it } from 'vitest'
import block from '#playground/BP/blocks/radio.block'

describe('define block', () => {
	it('returns a valid block object', () => {
		expect(block).toMatchSnapshot()
	})
})
