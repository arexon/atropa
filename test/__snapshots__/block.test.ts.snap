// Vitest Snapshot v1

exports[`define block > returns a valid block object 1`] = `
{
  "format_version": "1.19.10",
  "minecraft:block": {
    "components": {
      "minecraft:display_name": "test:radio",
      "minecraft:on_interact": {
        "condition": "q.is_sneaking",
        "event": "test:toggle",
      },
    },
    "description": {
      "identifier": "test:radio",
      "properties": {
        "test:is_on": [
          false,
          true,
        ],
      },
    },
    "events": {
      "test:toggle": {
        "set_block_property": {
          "test:is_on": "!q.block_property('test:is_on')",
        },
      },
    },
    "permutations": [
      {
        "components": {
          "minecraft:geometry": "geometry.radio.off",
        },
        "condition": "q.block_property('test:is_on') == false",
      },
      {
        "components": {
          "minecraft:geometry": "geometry.radio.on",
        },
        "condition": "q.block_property('test:is_on') == true",
      },
    ],
  },
}
`;
